src/main/kotlin/org/jitsi/nlj/rtp/ResumableStreamRewriter.kt

```diff
@@ -131,36 +103,151 @@ class ResumableStreamRewriter(
+                    /* First index seen. */
+                    insertItem(RewriteHistoryItem(accept, index), index)
+
+                    firstIndex = index
+
+                    newIndex = index
                 }
+
+                (index > lastIndex) -> {
+                    /* New.  Roll forward, filling in gap if necessary. */
+
+                    val newestIndex = lastIndex
+
+                    val newest = getContainer(newestIndex)
+                        ?: throw IllegalStateException("No newest container found")
+
+                    val newestNewIndex = newest.item!!.newIndex
+
+                    val indexGap = index - newestIndex
+
+                    val newGap = indexGap - if (accept) 0 else 1
+
+                    newIndex = newestNewIndex + newGap
+
+                    insertItem(RewriteHistoryItem(accept, newIndex), index)
+
+                    fillBetween(newestIndex + 1, index - 1, newestNewIndex + 1)
+                }
+
+                (index > lastIndex - size && index >= firstIndex) -> {
+                    /* In history.  Retrieve. */
+
+                    val container = getContainer(index)
+                        ?: throw IllegalStateException("No container found for index $index in history")
+                    val item = container.item!!
+
+                    if (item.accept == null) {
+                        item.accept = accept
+                        if (!accept)
+                            gapsLeft++
+                    }
+
+                    newIndex = item.newIndex
+                }
+
+                (index > lastIndex - size && index < firstIndex) -> {
+                    /* Older than previous oldest, but still in range of the map.
+                       Project map backwards. */
+
+                    val oldestIndex = Math.max(index - size, firstIndex)
+                    val oldest = getContainer(oldestIndex)
+                        ?: throw IllegalStateException("No oldest container found")
+
+                    val oldestNewIndex = oldest.item!!.newIndex
+
+                    val indexGap = index - oldestIndex /* Negative */
+
+                    val newGap = indexGap + if (accept) 0 else 1
+
+                    newIndex = oldestNewIndex + newGap
+
+                    insertItem(RewriteHistoryItem(accept, newIndex), index)
+
+                    fillBetween(index + 1, oldestIndex - 1, newIndex + 1)
+
+                    firstIndex = index
+                }
+
+                else -> {
+                    /* Older than the map. */
+
+                    val oldestIndex = Math.max(index - size, firstIndex)
+                    val oldest = getContainer(oldestIndex)
+                        ?: throw IllegalStateException("No oldest container found")
+                    val oldestDelta = oldestIndex - oldest.item!!.newIndex
+
+                    newIndex = index - oldestDelta
+                }
+            }
+
+            if (accept) return newIndex and 0xffff
```

Comment:
If I'm reading this right, we have `toPositiveInt` for this

Reply:
It's not making it positive, it's un-indexing it (mapping it back to the 16-bit range). Do we have a function for that?

Reply:
is that just `toShort()`?

Reply:
No, because it's still an `Int`, in the 0-65535 range.

Reply:
> in the 0-65535 range
I think we have `toPositiveShort()` as well.

Reply:
I think it would be `index.toShort().toPositiveInt()` but it's not clear to me that's clearer or more efficient.

Reply:
`and`ing with `0xfff` could imply a few things, which is why I think it'd be nice to have _some_ named function for it that's more descriptive about what we're trying to achieve...if the existing ones don't fit the bill we could define one in this file.